src/pykeen/models/unimodal/rgcn/model.py

```diff
@@ -0,0 +1,302 @@
+            # Get random dropout mask
+            edge_keep_mask = torch.rand(self.sources.shape[0], device=x.device) > self.edge_dropout
+
+            # Apply to edges
+            sources = sources[edge_keep_mask]
+            targets = targets[edge_keep_mask]
+            edge_types = edge_types[edge_keep_mask]
+
+        # Different dropout for self-loops (only in training mode)
+        if self.training and self.self_loop_dropout is not None:
+            node_keep_mask = torch.rand(x.shape[0], device=x.device) > self.self_loop_dropout
+        else:
+            node_keep_mask = None
+
+        # fixed edges -> pre-compute weights
+        if self.edge_weighting is not None:
+            edge_weights = torch.empty_like(sources, dtype=torch.float32)
+            for r in range(edge_types.max().item() + 1):
+                mask = edge_types == r
+                if mask.any():
+                    edge_weights[mask] = self.edge_weighting(sources[mask], targets[mask])
+        else:
+            edge_weights = None
+
+        for layer in self.layers:
+            if isinstance(layer, Decomposition):
+                kwargs = dict(
+                    node_keep_mask=node_keep_mask,
+                    source=sources,
+                    target=targets,
+                    edge_type=edge_types,
+                    edge_weights=edge_weights,
+                )
+            else:
+                kwargs = dict()
+            x = layer(x, **kwargs)
+
+        # Cache enriched representations
+        self.enriched_embeddings = x
+
+        if indices is not None:
+            x = x[indices]
+
+        return x
+
+
+class RGCN(
+    ERModel[torch.FloatTensor, RelationRepresentation, torch.FloatTensor],
+):
+    """An implementation of R-GCN from [schlichtkrull2018]_.
+
+    This model uses graph convolutions with relation-specific weights.
+
+    .. seealso::
+
+       - `Pytorch Geometric"s implementation of R-GCN
+         <https://github.com/rusty1s/pytorch_geometric/blob/1.3.2/examples/rgcn.py>`_
+       - `DGL"s implementation of R-GCN
+         <https://github.com/dmlc/dgl/tree/v0.4.0/examples/pytorch/rgcn>`_
+    """
+
+    #: The default strategy for optimizing the model"s hyper-parameters
+    hpo_default = dict(
+        embedding_dim=dict(type=int, low=32, high=512, q=32),
+        num_layers=dict(type=int, low=1, high=5, q=1),
+        use_bias=dict(type="bool"),
+        use_batch_norm=dict(type="bool"),
+        activation_cls=dict(type="categorical", choices=[nn.ReLU, nn.LeakyReLU]),
+        interaction=dict(type="categorical", choices=["distmult", "complex", "ermlp"]),
+        edge_dropout=dict(type=float, low=0.0, high=.9),
+        self_loop_dropout=dict(type=float, low=0.0, high=.9),
+        edge_weighting=dict(type="categorical", choices=["inverse_in_degree", "inverse_out_degree", "symmetric"]),
+        decomposition=dict(type="categorical", choices=["bases", "blocks"]),
+        # TODO: Decomposition kwargs
```

Comment:
@cthoyt how do you specify nested search spaces in optuna / the hpo_default?

Reply:
oh booyyyyyy. you should be able to add arbitrary logic inside the `pykeen.hpo.hpo._suggest_kwargs()` that checks based on the model type what you should do. I can look into this for you later